ENH Add "ensure_non_negative" option to check_array

TamaraAtanasoska opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes #29508.
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Adding an option to check if an array has `only_non_negative` values to the `sklearn.utils.validation.check_array` function, that contains the `sklearn.utils.validation.check_non_negative` functionality.

While in the initial issue I proposed `ensure_positive` as a name for the option, I found `only_non_negative` with exactly the same function in `sklearn.utils.validation._check_sample_weight`([here](https://github.com/scikit-learn/scikit-learn/blob/211a7f16a3b735c164dab5676115cb170353c327/sklearn/utils/validation.py#L2056)) and I used the same for consistency.

This addition will prevent the need to use `sklearn.utils.validation.check_non_negative` after `sklearn.utils.validation.check_array` in the use case of needing only non-negative values in an array. I am keeping the proposed changes minimal for easier review, but I am happy to also change all occurrences in the scikit-learn code where this pattern exists as an added commit.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

TamaraAtanasoska committed:
Add only_non_negative as check_array option

TamaraAtanasoska committed:
Fix docstring

jeremiedbb commented:
Thanks for the PR @TamaraAtanasoska. I'd like to keep the `ensure_xxx` naming pattern for consistency in `check_array`. Since `_check_sample_weight` is private, I'd rather change the name there if we really want to for both to have the same name.

It's possible though that ``ensure_positive`` may be confusing w.r.t whether or not zero is allowed. So `ensure_non_negative` is a valid option too.

TamaraAtanasoska committed:
Replace only_non_negative with ensure_non_negative

TamaraAtanasoska commented:
> Thanks for the PR @TamaraAtanasoska. I'd like to keep the `ensure_xxx` naming pattern for consistency in `check_array`. Since `_check_sample_weight` is private, I'd rather change the name there if we really want to for both to have the same name.
> 
> It's possible though that `ensure_positive` may be confusing w.r.t whether or not zero is allowed. So `ensure_non_negative` is a valid option too.

Agreed, changed it everywhere to `ensure_non_negative` in [e18e5f3](https://github.com/scikit-learn/scikit-learn/pull/29540/commits/e18e5f3986c685bb3fd45a8c2e44f5462972c5b5).

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -790,6 +791,10 @@ def check_array(
 
+    ensure_non_negative : bool, default=False
+        Make sure the array has only non-negative values. An array that contains
+        non-negative values will raise a ValueError.
```
```suggestion
        non-negative values will raise a ValueError.

        .. versionadded:: 1.6
```

glemaitre reviewed (commented):
We need to add an entry in the changelog located in `doc/whats_new/v1.6.rst` since it touch the public API of `check_array`. We don't need to document anything about `_check_sample_weight` there.

It is a bit weird that we introduce the option but that we don't use it.

@jeremiedbb do you think that we can use this feature directly in `NMF` because if I recall well this is one requirement of the transformer. I assume that we are currently calling the `check_non_negative` function directly in the code.

I don't know if there is something similar with transformer that expect counts as input in the feature extraction for text (like `TfidfTransformer`).

TamaraAtanasoska committed:
Add @glemaitre's versioning suggestion in docstring

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

TamaraAtanasoska commented:
> doc/whats_new/v1.6.rst

Thanks for the review @glemaitre! I am looking into adding to the changelog now. Is this an `enhancement` or a `feature`? I will also add it in the places where `check_non_negative` follows `check_array`.

TamaraAtanasoska committed:
Relocate check within check_array out of neseted if

TamaraAtanasoska committed:
Example usage of ensure_non_negative

glemaitre commented:
>  Is this an enhancement or a feature?

I would say an enhancement.

TamaraAtanasoska commented:
@glemaitre @jeremiedbb a few notes on [a6cfa3a](https://github.com/scikit-learn/scikit-learn/pull/29540/commits/a6cfa3ad46bf9af7ad399d755eaf2238e0e6c3db):
- I added a short changelog entry. Let me know if it needs to be changed in any way, first time 😊 
- I had to move the check at the very bottom of `check_array`, which mirrors its usage until now. Otherwise the input array is not "ready" and some changes will need to be made to `check_non_negative`.
- I added two example usages in the existing code. What comes to light when looking at the changes required for them is that there is a loss of information that previous was there, packed in `whom` which is not a generic `check_array` source of input. In what cases is this information essential?

TamaraAtanasoska committed:
Fix doc error

TamaraAtanasoska committed:
Fix references in docsstring

glemaitre reviewed (commented):
> I added two example usages in the existing code. What comes to light when looking at the changes required for them is that there is a loss of information that previous was there, packed in whom which is now a generic check_array as source of input. In what cases is this information essential?

I think that we can alleviate this issue by using both `input_name` and `estimator` of the `check_array`:  `estimator` will help at getting the name of the estimator while `input_name` will help at indicating which data container contains negative values. We could adapt a bit the error message of `check_non_negative` to have a meaningful message, or craft a good `whom` message from these two parameters.

glemaitre commented on doc/whats_new/v1.6.rst:
```diff
@@ -264,6 +264,14 @@ Changelog
+:mod:`sklearn.utils`
+....................
+
+- |Enhancement| :func:`utils.validation.check_array` now accepts `ensure_non_negative` to
+  check for negative values in the passed array, until now only available through calling
+  :func:`utils.validation.check_non_negative`.
+  :pr:`29540` by :user:`Tamara Atanasoska <tamaraatanasoska>`.
```
```suggestion
- |Enhancement| :func:`utils.validation.check_array` now accepts `ensure_non_negative`
  to check for negative values in the passed array, until now only available through
  calling :func:`utils.validation.check_non_negative`.
  :pr:`29540` by :user:`Tamara Atanasoska <tamaraatanasoska>`.
```

glemaitre replied:
Just to respect the 88 characters :)

Otherwise this is good entry.

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -1120,6 +1127,9 @@ def is_sparse(dtype):
                 array = array.copy(**copy_params)
 
+    if ensure_non_negative:
+        check_non_negative(array, "check_array")
```
As stated in another comment, if `input_name` and `estimator` are not the default, I think that we can make a crafted message better than `"check_array"`.

glemaitre reviewed (commented):
I think that we can also modify `NMF`. Right now the `check_non_negative` is called in ` _fit_transform` method. However, we could remove it and use the `ensure_non_negative` of `check_array` call in `fit_transform`  and `transform`

TamaraAtanasoska committed:
Update doc/whats_new/v1.6.rst

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

TamaraAtanasoska committed:
Add a meaningful error message

TamaraAtanasoska commented:
@glemaitre I tired to address all the comments above in [890b480](https://github.com/scikit-learn/scikit-learn/pull/29540/commits/890b4804f04a0eca0da0ddb0097ae6bebda57bf7). Not sure about adding a generic `"estimator"` in [this line](https://github.com/scikit-learn/scikit-learn/pull/29540/commits/890b4804f04a0eca0da0ddb0097ae6bebda57bf7#diff-962bea3f949b0a1e7b2ad40d2879f5d7ba1e6cd7da168daa82f84a87e800715eR1131). Would it be better if it was empty maybe?

glemaitre commented:
> Would it be better if it was empty maybe?

I would probably not have a generic name and not mentioned it (so probably the empty yes). It would translate to something like:

```python
    if ensure_non_negative:
        whom = input_name
        if estimator_name:
            whom += f" in {estimator_name}"
        check_non_negative(array, whom=whom)
```

TamaraAtanasoska committed:
Reformat error message

TamaraAtanasoska commented:
> ```python
> whom = input_name
>         if estimator_name:
>             whom += f" in {estimator_name}"
> ```

I love this suggestion, just pushed a commit with it. It preserves the `whom` that makes the usage of it obvious and adds a more natural end of sentence without an estimator name.

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -790,6 +791,12 @@ def check_array(
            Accepts `pd.NA` and converts it into `np.nan`
 
+    ensure_non_negative : bool, default=False
+        Make sure the array has only non-negative values. An array that contains
+        non-negative values will raise a ValueError.
```
```suggestion
        Make sure the array has only non-negative values. If True, an array that contains
        negative values will raise a ValueError.
```

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -1120,6 +1127,12 @@ def is_sparse(dtype):
             else:
                 array = array.copy(**copy_params)
 
+    if ensure_non_negative:
+        whom = input_name
+        if estimator_name:
+            whom += f" in {estimator_name}"
+        check_non_negative(array, whom)
+
```
I would put that just before the previous ``if force_writeable`` block to be able to fail fast, before a pontential copy.

jeremiedbb commented on sklearn/neighbors/_base.py:
```diff
@@ -178,8 +177,9 @@ def _check_precomputed(X):
     copied = graph.format != "csr"
-    graph = check_array(graph, accept_sparse="csr")
-    check_non_negative(graph, whom="precomputed distance matrix.")
+    graph = check_array(
+        graph, accept_sparse="csr", ensure_non_negative=True, input_name="graph"
+    )
```
why didn't you keep `precomputed distance matrix` ?

glemaitre replied:
Agreed, `graph` is an internal variable and people might get confused by the error message then.

TamaraAtanasoska replied:
thanks @jeremiedbb , makes sense! I didn't notice there is more relevant naming there.

jeremiedbb commented:
@glemaitre, should we add it to ``check_X_y`` as well ? it has pretty much the same parameters as `check_array`, although I don't think we ever use it in conjonction with ``check_non_negative``.

jeremiedbb committed:
iter

jeremiedbb reviewed (approved):
LGTM

glemaitre commented:
> @glemaitre, should we add it to check_X_y as well ? it has pretty much the same parameters as check_array, although I don't think we ever use it in conjonction with check_non_negative.

Yep we can do that but in a subsequent PR.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/TamaraAtanasoska/29540

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre reviewed (approved):
OK after failing at properly using `git merge`, I think this is all good ;)

glemaitre commented:
Enabling auto-merge. Thanks @TamaraAtanasoska
